File: python/dssatutils/weather_era5land.py
# File: weather_era5land.py
# Python port of weather_era5land.R
#
# Downloads ERA5-Land point time series data from Copernicus Climate Data Store (CDS)
# in CSV format, aggregates it to daily statistics, and writes DSSAT-formatted .WTH files.
#
# CDS dataset: "reanalysis-era5-land-timeseries"

import os
import re
import gc
import datetime
from datetime import date, timedelta
import numpy as np
import pandas as pd
from concurrent.futures import ThreadPoolExecutor

from .credentials import make_cds_client

# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _log_worker_message(log_file, level="INFO", point_id=None, msg=""):
    if not log_file:
        return
    os.makedirs(os.path.dirname(log_file), exist_ok=True)
    id_part = f" [ID={point_id}]" if point_id else ""
    timestamp = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    clean_msg = " ".join(str(msg).split())
    line = f"[{timestamp}] [{level}] [WEATHER_ERA5_LAND]{id_part} {clean_msg}\n"
    with open(log_file, "a") as f:
        f.write(line)


def _fill_na_with_neighbor_mean(x, window=2, min_neighbors=2, max_iter=10):
    out = np.array(x, dtype=float)
    if not np.isnan(out).any():
        return out
    n = len(out)
    for _ in range(max_iter):
        missing_idx = np.where(np.isnan(out))[0]
        if len(missing_idx) == 0:
            break
        prev_out = out.copy()
        for idx in missing_idx:
            start_i = max(0, idx - window)
            end_i = min(n, idx + window + 1)
            neighbor_vals = [prev_out[i] for i in range(start_i, end_i) if i != idx]
            neighbor_vals = [v for v in neighbor_vals if not np.isnan(v)]
            if len(neighbor_vals) >= min_neighbors:
                out[idx] = sum(neighbor_vals) / len(neighbor_vals)
        if np.array_equal(np.isnan(out), np.isnan(prev_out)):
            break
    return out


def _find_first_matching_column(cols, patterns):
    for pattern in patterns:
        for col in cols:
            if re.search(pattern, col, re.IGNORECASE):
                return col
    return None


def _parse_hourly_datetime(df):
    cols = df.columns
    date_col = _find_first_matching_column(cols, [r"^date$", r"date", r"valid_date"])
    time_col = _find_first_matching_column(cols, [r"^time$", r"hour", r"valid_time"])
    dt_col = _find_first_matching_column(cols, [r"datetime", r"date_time", r"valid_datetime", r"^timestamp$", r"^time$"])

    if dt_col is not None:
        try:
            return pd.to_datetime(df[dt_col], utc=True)
        except Exception:
            pass

    if date_col is not None and time_col is not None:
        try:
            return pd.to_datetime(df[date_col].astype(str) + " " + df[time_col].astype(str), utc=True)
        except Exception:
            pass

    if date_col is not None:
        try:
            return pd.to_datetime(df[date_col], utc=True)
        except Exception:
            pass

    raise ValueError("Could not identify a datetime column in the ERA5-Land CSV download.")


def _calc_rh_from_temp_dew(temp_c, dew_c):
    rh = 100 * np.exp((17.625 * dew_c) / (243.04 + dew_c) - (17.625 * temp_c) / (243.04 + temp_c))
    return np.clip(rh, 0.0, 100.0)


def _cap_end_date_era5_land(end_year, lag_days=5):
    requested_end = date(end_year, 12, 31)
    safe_end = date.today() - timedelta(days=lag_days)
    return min(requested_end, safe_end)


def _download_era5_land_point_csv(latitude, longitude, start_date_str, end_date_str, target_file, cds_user="ecmwfr"):
    try:
        import cdsapi  # imported lazily
    except ImportError as exc:
        raise ImportError("ERA5-Land requires the 'cdsapi' Python package.") from exc
    c = make_cds_client(cdsapi)

    req = {
        'variable': [
            '2m_temperature',
            '2m_dewpoint_temperature',
            'total_precipitation',
            'surface_solar_radiation_downwards',
            '10m_u_component_of_wind',
            '10m_v_component_of_wind'
        ],
        'location': {
            'latitude': float(latitude),
            'longitude': float(longitude)
        },
        'date': {
            'start': start_date_str,
            'end': end_date_str
        },
        'data_format': 'csv'
    }

    os.makedirs(os.path.dirname(target_file), exist_ok=True)
    c.retrieve('reanalysis-era5-land-timeseries', req, target_file)


def _read_era5_land_hourly_csv(csv_file):
    df = pd.read_csv(csv_file)
    if df.empty:
        raise ValueError(f"Downloaded ERA5-Land file is empty: {csv_file}")

    cols = df.columns
    time_values = _parse_hourly_datetime(df)

    temp_col = _find_first_matching_column(cols, [r"^2m_temperature$", r"2m_temperature"])
    dew_col  = _find_first_matching_column(cols, [r"^2m_dewpoint_temperature$", r"2m_dewpoint_temperature"])
    tp_col   = _find_first_matching_column(cols, [r"^total_precipitation$", r"total_precipitation"])
    ssrd_col = _find_first_matching_column(cols, [r"^surface_solar_radiation_downwards$", r"surface_solar_radiation_downwards"])
    u10_col  = _find_first_matching_column(cols, [r"^10m_u_component_of_wind$", r"10m_u_component_of_wind"])
    v10_col  = _find_first_matching_column(cols, [r"^10m_v_component_of_wind$", r"10m_v_component_of_wind"])

    needed = [temp_col, dew_col, tp_col, ssrd_col, u10_col, v10_col]
    if any(c is None for c in needed):
        found = ", ".join(cols)
        raise ValueError(f"Could not find all required ERA5-Land columns in {csv_file}. Found: {found}")

    return pd.DataFrame({
        'DATETIME_UTC': time_values,
        'T2M_K': pd.to_numeric(df[temp_col], errors='coerce'),
        'DEW_K': pd.to_numeric(df[dew_col], errors='coerce'),
        'TP_M': pd.to_numeric(df[tp_col], errors='coerce'),
        'SSRD_J': pd.to_numeric(df[ssrd_col], errors='coerce'),
        'U10': pd.to_numeric(df[u10_col], errors='coerce'),
        'V10': pd.to_numeric(df[v10_col], errors='coerce'),
    })


def _aggregate_era5_land_to_daily(hourly_df, start_date_str, end_date_str, utc_offset_hours=None):
    if utc_offset_hours is not None:
        hourly_df["DATETIME_LOCAL"] = hourly_df["DATETIME_UTC"] + pd.to_timedelta(float(utc_offset_hours), unit="h")
    else:
        hourly_df["DATETIME_LOCAL"] = hourly_df["DATETIME_UTC"]

    hourly_df["DATE_obj"] = hourly_df["DATETIME_LOCAL"].dt.date
    hourly_df["T2M_C"] = hourly_df["T2M_K"] - 273.15
    hourly_df["DEW_C"] = hourly_df["DEW_K"] - 273.15
    hourly_df["RAIN_MM"] = np.maximum(hourly_df["TP_M"], 0.0) * 1000.0
    hourly_df["SRAD_MJ"] = np.maximum(hourly_df["SSRD_J"], 0.0) / 1e6
    hourly_df["WIND_MS"] = np.sqrt(hourly_df["U10"]**2 + hourly_df["V10"]**2)
    hourly_df["RH2M"] = _calc_rh_from_temp_dew(hourly_df["T2M_C"], hourly_df["DEW_C"])

    daily = hourly_df.groupby("DATE_obj").agg(
        SRAD=("SRAD_MJ", "sum"),
        TMAX=("T2M_C", "max"),
        TMIN=("T2M_C", "min"),
        RAIN=("RAIN_MM", "sum"),
        TDEW=("DEW_C", "mean"),
        RH2M=("RH2M", "mean"),
        WIND=("WIND_MS", "mean")
    ).reset_index()

    start_d = pd.to_datetime(start_date_str).date()
    end_d = pd.to_datetime(end_date_str).date()

    full_dates = pd.date_range(start_d, end_d, freq="D").date
    full_calendar = pd.DataFrame({"DATE_obj": full_dates})
    weather_data = pd.merge(full_calendar, daily, on="DATE_obj", how="left")

    vars_to_repair = ["SRAD", "TMAX", "TMIN", "RAIN", "TDEW", "RH2M", "WIND"]
    missing_before = {v: int(weather_data[v].isna().sum()) for v in vars_to_repair}

    for v in vars_to_repair:
        weather_data[v] = _fill_na_with_neighbor_mean(weather_data[v].values)

    weather_data["RAIN"] = np.maximum(weather_data["RAIN"], 0.0)
    weather_data["RH2M"] = np.clip(weather_data["RH2M"], 0.0, 100.0)

    bad_temp_idx = weather_data["TMAX"] < weather_data["TMIN"]
    if bad_temp_idx.any():
        tmp = weather_data.loc[bad_temp_idx, "TMAX"].copy()
        weather_data.loc[bad_temp_idx, "TMAX"] = weather_data.loc[bad_temp_idx, "TMIN"]
        weather_data.loc[bad_temp_idx, "TMIN"] = tmp

    missing_after = {v: int(np.isnan(weather_data[v].values).sum()) for v in vars_to_repair}

    weather_data["DATE_obj"] = pd.to_datetime(weather_data["DATE_obj"])
    weather_data["YEAR"] = weather_data["DATE_obj"].dt.year
    weather_data["MM"] = weather_data["DATE_obj"].dt.month
    weather_data["DOY"] = weather_data["DATE_obj"].dt.dayofyear

    return weather_data, missing_before, missing_after


def _write_dssat_weather_file(weather_data, latitude, longitude, output_file, point_id):
    weather_data["TAVG"] = (weather_data["TMAX"] + weather_data["TMIN"]) / 2.0
    tav = float(weather_data["TAVG"].mean())

    monthly_temps = weather_data.groupby(["YEAR", "MM"])["TAVG"].mean().reset_index()
    annual_amps = monthly_temps.groupby("YEAR")["TAVG"].agg(lambda x: x.max() - x.min()).reset_index()
    amp = float(annual_amps["TAVG"].mean())

    lines = [
        f"$WEATHER DATA: ERA5-LAND  (Point ID: {point_id})",
        "@ INSI      LAT     LONG  ELEV   TAV   AMP REFHT WNDHT",
        f"E5LD    {latitude:8.4f} {longitude:8.4f}   -99 {tav:5.1f} {amp:5.1f}   -99   -99",
        "@  DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND"
    ]

    for _, row in weather_data.iterrows():
        date_str = f"{int(row['YEAR'])}{int(row['DOY']):03d}"
        srad = -99.0 if pd.isna(row['SRAD']) else float(row['SRAD'])
        tmax = -99.0 if pd.isna(row['TMAX']) else float(row['TMAX'])
        tmin = -99.0 if pd.isna(row['TMIN']) else float(row['TMIN'])
        rain = -99.0 if pd.isna(row['RAIN']) else float(row['RAIN'])
        tdew = -99.0 if pd.isna(row['TDEW']) else float(row['TDEW'])
        rh2m = -99.0 if pd.isna(row['RH2M']) else float(row['RH2M'])
        wind = -99.0 if pd.isna(row['WIND']) else float(row['WIND'])

        line = (
            f"{date_str:>7s}"
            f"{srad:6.1f}"
            f"{tmax:6.1f}"
            f"{tmin:6.1f}"
            f"{rain:6.1f}"
            f"{tdew:6.1f}"
            f"{rh2m:6.1f}"
            f"{wind:6.1f}"
        )
        lines.append(line)

    with open(output_file, "w") as f:
        f.write("\n".join(lines) + "\n")


# ---------------------------------------------------------------------------
# Public entry point
# ---------------------------------------------------------------------------

def process_weather_era5_land(
    shapefile,
    start_year: int,
    end_year: int,
    output_dir: str,
    id_col: str,
    lat_col: str,
    lon_col: str,
    n_cores: int,
    log_file: str,
    cds_user: str = "ecmwfr",
    utc_offset_hours = None,
    cache_dir: str = None,
    keep_raw_downloads: bool = False,
    availability_lag_days: int = 5
) -> None:
    """Download ERA5-Land point time series data and write DSSAT .WTH files.

    Requires cdsapi installed and configured with ~/.cdsapirc API key.
    """
    print(f"--- Starting ERA5-Land Download (Years: {start_year}–{end_year}) ---")

    start_date_str = f"{start_year}-01-01"
    end_date = _cap_end_date_era5_land(end_year, lag_days=availability_lag_days)
    end_date_str = str(end_date)

    os.makedirs(output_dir, exist_ok=True)
    if cache_dir is None:
        cache_dir = os.path.join(output_dir, "_era5_cache")
    os.makedirs(cache_dir, exist_ok=True)

    jobs = []
    for _, row in shapefile.iterrows():
        point_id = str(row[id_col])
        lat = float(row[lat_col])
        lon = float(row[lon_col])
        out_file = os.path.join(output_dir, f"{point_id}.WTH")
        raw_csv = os.path.join(cache_dir, f"{point_id}_era5land_hourly.csv")
        if not os.path.exists(out_file):
            jobs.append((point_id, lat, lon, out_file, raw_csv))

    if not jobs:
        print("All ERA5-Land weather files already exist. Processing skipped.")
        return

    def _process_one(job):
        point_id, lat, lon, out_file, raw_csv = job
        try:
            _download_era5_land_point_csv(lat, lon, start_date_str, end_date_str, raw_csv, cds_user)
            hourly = _read_era5_land_hourly_csv(raw_csv)
            daily, missing_before, missing_after = _aggregate_era5_land_to_daily(
                hourly, start_date_str, date_str := end_date_str, utc_offset_hours
            )

            if any(val > 0 for val in missing_before.values()):
                repair_msg = "; ".join(f"{v} {missing_before[v]}->{missing_after[v]}" for v in missing_before)
                _log_worker_message(log_file, "WARN", point_id,
                                    f"Missing ERA5-Land daily values detected and repaired: {repair_msg}")

            _write_dssat_weather_file(daily, lat, lon, out_file, point_id)

            if not keep_raw_downloads and os.path.exists(raw_csv):
                os.remove(raw_csv)
            _log_worker_message(log_file, "INFO", point_id, f"Successfully created ERA5-Land weather file: {os.path.basename(out_file)}")
        except Exception as e:
            _log_worker_message(log_file, "ERROR", point_id, str(e))
            print(f"Point {point_id} failed: {e}")

    requested_cores = max(1, int(n_cores))
    if requested_cores == 1 or len(jobs) == 1:
        for job in jobs:
            _process_one(job)
    else:
        with ThreadPoolExecutor(max_workers=min(requested_cores, len(jobs))) as executor:
            list(executor.map(_process_one, jobs))

    print(f"ERA5-Land weather processing complete. Output: {output_dir}")

File: R/weather_era5land.R
# File: R/weather_era5land.R
# Centralized ERA5-Land weather downloader and parser for dssatutils.

.log_worker_message <- function(log_file, level = "INFO", point_id = NULL, msg = "") {
  if (is.null(log_file) || !nzchar(log_file)) return(invisible(NULL))
  dir.create(dirname(log_file), recursive = TRUE, showWarnings = FALSE)
  id_part <- if (!is.null(point_id) && !is.na(point_id) && nzchar(as.character(point_id))) {
    sprintf(" [ID=%s]", as.character(point_id))
  } else {
    ""
  }
  line <- sprintf(
    "[%s] [%s] [WEATHER_ERA5_LAND]%s %s",
    format(Sys.time(), "%Y-%m-%d %H:%M:%S %Z"),
    level,
    id_part,
    gsub("[\r\n\t]+", " ", paste(as.character(msg), collapse = " "))
  )
  cat(line, "\n", file = log_file, append = TRUE)
  invisible(NULL)
}

.fill_na_with_neighbor_mean <- function(x, window = 2, min_neighbors = 2, max_iter = 10) {
  if (!is.numeric(x)) return(x)

  out <- as.numeric(x)
  if (!anyNA(out)) return(out)

  n <- length(out)

  for (iter in seq_len(max_iter)) {
    missing_idx <- which(is.na(out))
    if (length(missing_idx) == 0) break

    prev_out <- out

    for (idx in missing_idx) {
      neighbor_idx <- setdiff(seq.int(max(1, idx - window), min(n, idx + window)), idx)
      neighbor_vals <- prev_out[neighbor_idx]
      neighbor_vals <- neighbor_vals[!is.na(neighbor_vals)]
      if (length(neighbor_vals) >= min_neighbors) {
        out[idx] <- mean(neighbor_vals)
      }
    }

    if (identical(which(is.na(out)), which(is.na(prev_out)))) break
  }

  out
}

.find_first_matching_column <- function(nm, patterns) {
  if (length(nm) == 0) return(NA_character_)
  for (pat in patterns) {
    idx <- grep(pat, nm, ignore.case = TRUE, perl = TRUE)
    if (length(idx) > 0) return(nm[idx[1]])
  }
  NA_character_
}

.parse_hourly_datetime <- function(df) {
  nm <- names(df)

  date_col <- .find_first_matching_column(nm, c("^date$", "date", "valid_date"))
  time_col <- .find_first_matching_column(nm, c("^time$", "hour", "valid_time"))
  dt_col <- .find_first_matching_column(nm, c("datetime", "date_time", "valid_datetime", "^timestamp$", "^time$"))

  if (!is.na(dt_col)) {
    raw <- df[[dt_col]]
    parsed <- suppressWarnings(as.POSIXct(raw, tz = "UTC"))
    if (all(is.na(parsed))) parsed <- suppressWarnings(as.POSIXct(raw, tz = "UTC", format = "%Y-%m-%d %H:%M:%S"))
    if (all(is.na(parsed))) parsed <- suppressWarnings(as.POSIXct(raw, tz = "UTC", format = "%Y-%m-%dT%H:%M:%S"))
    if (!all(is.na(parsed))) return(parsed)
  }

  if (!is.na(date_col) && !is.na(time_col)) {
    raw <- paste(df[[date_col]], df[[time_col]])
    parsed <- suppressWarnings(as.POSIXct(raw, tz = "UTC"))
    if (all(is.na(parsed))) parsed <- suppressWarnings(as.POSIXct(raw, tz = "UTC", format = "%Y-%m-%d %H:%M:%S"))
    if (all(is.na(parsed))) parsed <- suppressWarnings(as.POSIXct(raw, tz = "UTC", format = "%Y-%m-%d %H:%M"))
    if (!all(is.na(parsed))) return(parsed)
  }

  if (!is.na(date_col)) {
    parsed <- suppressWarnings(as.POSIXct(df[[date_col]], tz = "UTC"))
    if (!all(is.na(parsed))) return(parsed)
  }

  stop("Could not identify a datetime column in the ERA5-Land CSV download.")
}

.calc_rh_from_temp_dew <- function(temp_c, dew_c) {
  rh <- 100 * exp((17.625 * dew_c) / (243.04 + dew_c) - (17.625 * temp_c) / (243.04 + temp_c))
  pmin(pmax(rh, 0), 100)
}

.ensure_numeric <- function(x) {
  suppressWarnings(as.numeric(x))
}

.cap_end_date_era5_land <- function(end_year, lag_days = 5) {
  requested_end <- as.Date(sprintf("%d-12-31", end_year))
  min(requested_end, Sys.Date() - lag_days)
}

.build_era5_land_request <- function(latitude,
                                     longitude,
                                     start_date,
                                     end_date,
                                     target_file,
                                     data_format = "csv",
                                     variables = c(
                                       "2m_temperature",
                                       "2m_dewpoint_temperature",
                                       "total_precipitation",
                                       "surface_solar_radiation_downwards",
                                       "10m_u_component_of_wind",
                                       "10m_v_component_of_wind"
                                     )) {
  if (!nzchar(target_file)) stop("target_file must be supplied.")
  if (!(data_format %in% c("csv", "netcdf"))) stop("data_format must be 'csv' or 'netcdf'.")

  list(
    dataset_short_name = "reanalysis-era5-land-timeseries",
    variable = variables,
    location = list(
      latitude = as.numeric(latitude),
      longitude = as.numeric(longitude)
    ),
    date = list(
      start = as.character(start_date),
      end = as.character(end_date)
    ),
    data_format = data_format,
    target = basename(target_file)
  )
}

.download_era5_land_point_csv <- function(latitude,
                                          longitude,
                                          start_date,
                                          end_date,
                                          target_file,
                                          cds_user = "ecmwfr",
                                          verbose = FALSE) {
  dir.create(dirname(target_file), recursive = TRUE, showWarnings = FALSE)
  req <- .build_era5_land_request(
    latitude = latitude,
    longitude = longitude,
    start_date = start_date,
    end_date = end_date,
    target_file = target_file,
    data_format = "csv"
  )
  ecmwfr::wf_request(
    request = req,
    user = cds_user,
    transfer = TRUE,
    path = dirname(target_file),
    verbose = verbose
  )
}

.read_era5_land_hourly_csv <- function(csv_file) {
  df <- readr::read_csv(csv_file, show_col_types = FALSE, progress = FALSE, name_repair = "minimal")
  if (nrow(df) == 0) stop(sprintf("Downloaded ERA5-Land file is empty: %s", csv_file))

  nm <- names(df)
  time_values <- .parse_hourly_datetime(df)

  temp_col <- .find_first_matching_column(nm, c("^2m_temperature$", "2m_temperature"))
  dew_col  <- .find_first_matching_column(nm, c("^2m_dewpoint_temperature$", "2m_dewpoint_temperature"))
  tp_col   <- .find_first_matching_column(nm, c("^total_precipitation$", "total_precipitation"))
  ssrd_col <- .find_first_matching_column(nm, c("^surface_solar_radiation_downwards$", "surface_solar_radiation_downwards"))
  u10_col  <- .find_first_matching_column(nm, c("^10m_u_component_of_wind$", "10m_u_component_of_wind"))
  v10_col  <- .find_first_matching_column(nm, c("^10m_v_component_of_wind$", "10m_v_component_of_wind"))

  needed <- c(temp_col, dew_col, tp_col, ssrd_col, u10_col, v10_col)
  if (any(is.na(needed))) {
    stop(sprintf(
      "Could not find all required ERA5-Land columns in %s. Found names: %s",
      csv_file,
      paste(nm, collapse = ", ")
    ))
  }

  data.frame(
    DATETIME_UTC = time_values,
    T2M_K = .ensure_numeric(df[[temp_col]]),
    DEW_K = .ensure_numeric(df[[dew_col]]),
    TP_M = .ensure_numeric(df[[tp_col]]),
    SSRD_J = .ensure_numeric(df[[ssrd_col]]),
    U10 = .ensure_numeric(df[[u10_col]]),
    V10 = .ensure_numeric(df[[v10_col]]),
    stringsAsFactors = FALSE
  )
}

.aggregate_era5_land_to_daily <- function(hourly_df,
                                          start_date,
                                          end_date,
                                          utc_offset_hours = NULL) {
  if (!is.null(utc_offset_hours) && !is.na(utc_offset_hours)) {
    hourly_df$DATETIME_LOCAL <- hourly_df$DATETIME_UTC + lubridate::hours(as.numeric(utc_offset_hours))
  } else {
    hourly_df$DATETIME_LOCAL <- hourly_df$DATETIME_UTC
  }

  hourly_df$DATE_obj <- as.Date(hourly_df$DATETIME_LOCAL)
  hourly_df$T2M_C <- hourly_df$T2M_K - 273.15
  hourly_df$DEW_C <- hourly_df$DEW_K - 273.15
  hourly_df$RAIN_MM <- pmax(hourly_df$TP_M, 0) * 1000
  hourly_df$SRAD_MJ <- pmax(hourly_df$SSRD_J, 0) / 1e6
  hourly_df$WIND_MS <- sqrt(hourly_df$U10^2 + hourly_df$V10^2)
  hourly_df$RH2M <- .calc_rh_from_temp_dew(hourly_df$T2M_C, hourly_df$DEW_C)

  # Avoid using %>% inside parallel workers if namespace isn't fully attached
  daily <- hourly_df %>%
    dplyr::group_by(DATE_obj) %>%
    dplyr::summarise(
      SRAD = sum(SRAD_MJ, na.rm = TRUE),
      TMAX = max(T2M_C, na.rm = TRUE),
      TMIN = min(T2M_C, na.rm = TRUE),
      RAIN = sum(RAIN_MM, na.rm = TRUE),
      TDEW = mean(DEW_C, na.rm = TRUE),
      RH2M = mean(RH2M, na.rm = TRUE),
      WIND = mean(WIND_MS, na.rm = TRUE),
      .groups = "drop"
    ) %>%
    dplyr::arrange(DATE_obj)

  full_calendar <- data.frame(DATE_obj = seq(as.Date(start_date), as.Date(end_date), by = "day"))
  weather_data <- full_calendar %>%
    dplyr::left_join(daily, by = "DATE_obj") %>%
    dplyr::mutate(
      YEAR = lubridate::year(DATE_obj),
      MM = lubridate::month(DATE_obj),
      DOY = lubridate::yday(DATE_obj),
      DATE = sprintf("%d%03d", YEAR, DOY)
    )

  vars_to_repair <- c("SRAD", "TMAX", "TMIN", "RAIN", "TDEW", "RH2M", "WIND")
  missing_before <- vapply(weather_data[vars_to_repair], function(x) sum(is.na(x)), integer(1))

  for (v in vars_to_repair) {
    weather_data[[v]] <- .fill_na_with_neighbor_mean(weather_data[[v]], window = 2, min_neighbors = 2)
  }

  weather_data$RAIN[!is.na(weather_data$RAIN) & weather_data$RAIN < 0] <- 0
  weather_data$RH2M[!is.na(weather_data$RH2M)] <- pmin(pmax(weather_data$RH2M[!is.na(weather_data$RH2M)], 0), 100)

  bad_temp_order_idx <- which(!is.na(weather_data$TMAX) & !is.na(weather_data$TMIN) & weather_data$TMAX < weather_data$TMIN)
  if (length(bad_temp_order_idx) > 0) {
    tmp <- weather_data$TMAX[bad_temp_order_idx]
    weather_data$TMAX[bad_temp_order_idx] <- weather_data$TMIN[bad_temp_order_idx]
    weather_data$TMIN[bad_temp_order_idx] <- tmp
  }

  missing_after <- vapply(weather_data[vars_to_repair], function(x) sum(is.na(x)), integer(1))
  attr(weather_data, "missing_before") <- missing_before
  attr(weather_data, "missing_after") <- missing_after
  weather_data
}

.write_dssat_weather_file <- function(weather_data,
                                      latitude,
                                      longitude,
                                      output_file,
                                      point_id) {
  weather_data$TAVG <- (weather_data$TMAX + weather_data$TMIN) / 2
  tav <- mean(weather_data$TAVG, na.rm = TRUE)

  monthly_temps <- weather_data %>%
    dplyr::group_by(YEAR, MM) %>%
    dplyr::summarise(TAVG_MON = mean(TAVG, na.rm = TRUE), .groups = "drop")

  annual_amps <- monthly_temps %>%
    dplyr::group_by(YEAR) %>%
    dplyr::summarise(AMP_YR = max(TAVG_MON, na.rm = TRUE) - min(TAVG_MON, na.rm = TRUE), .groups = "drop")

  amp <- mean(annual_amps$AMP_YR, na.rm = TRUE)

  weather_data_out <- weather_data %>%
    dplyr::mutate(
      dplyr::across(c(SRAD, TMAX, TMIN, RAIN, TDEW, RH2M, WIND), ~ ifelse(is.na(.), -99, .)),
      DATE = sprintf("%d%03d", YEAR, DOY)
    )

  lines <- c(
    sprintf("$WEATHER DATA: ERA5-LAND  (Point ID: %s)", point_id),
    "@ INSI      LAT     LONG  ELEV   TAV   AMP REFHT WNDHT",
    sprintf("E5LD    %8.4f %8.4f   -99 %5.1f %5.1f   -99   -99", latitude, longitude, tav, amp),
    "@  DATE  SRAD  TMAX  TMIN  RAIN  TDEW  RH2M  WIND"
  )

  body <- with(
    weather_data_out,
    sprintf("%7s%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f%6.1f", DATE, SRAD, TMAX, TMIN, RAIN, TDEW, RH2M, WIND)
  )

  writeLines(c(lines, body), con = output_file)
  invisible(output_file)
}

#' Download ERA5-Land weather and write DSSAT .WTH files for all points
#'
#' @param shapefile sf object or data frame with point coordinates.
#' @param start_year Start year.
#' @param end_year End year.
#' @param output_dir Output directory for .WTH files.
#' @param id_col Point id column name.
#' @param lat_col Latitude column name.
#' @param lon_col Longitude column name.
#' @param n_cores Number of parallel workers.
#' @param log_file Log file path.
#' @param cds_user Keyring user entry for ecmwfr.
#' @param utc_offset_hours Optional fixed offset applied before daily aggregation.
#' @param cache_dir Directory for raw ERA5 downloads.
#' @param keep_raw_downloads Keep raw hourly CSVs.
#' @param availability_lag_days Conservative lag for latest available data. Defaults to 5.
#' @export
process_weather_era5_land <- function(shapefile,
                                      start_year,
                                      end_year,
                                      output_dir,
                                      id_col,
                                      lat_col,
                                      lon_col,
                                      n_cores,
                                      log_file,
                                      cds_user = "ecmwfr",
                                      utc_offset_hours = NULL,
                                      cache_dir = file.path(output_dir, "_era5_cache"),
                                      keep_raw_downloads = FALSE,
                                      availability_lag_days = 5) {
  .dssatutils_ensure_cds_credentials(user = cds_user, require_ecmwfr = TRUE)
  
  message(sprintf("--- Starting ERA5-Land Download (Years: %d-%d) ---", start_year, end_year))

  start_date_str <- paste0(start_year, "-01-01")
  end_date <- .cap_end_date_era5_land(end_year, lag_days = availability_lag_days)
  end_date_str <- as.character(end_date)

  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  dir.create(cache_dir, recursive = TRUE, showWarnings = FALSE)

  coords_list <- .extract_coords(shapefile, id_col, lat_col, lon_col)
  ids <- coords_list$ids
  lats <- coords_list$lats
  lons <- coords_list$lons

  .process_one <- function(i) {
    latitude <- lats[i]
    longitude <- lons[i]
    point_id <- ids[i]
    output_file <- file.path(output_dir, sprintf("%s.WTH", point_id))
    raw_csv <- file.path(cache_dir, sprintf("%s_era5land_hourly.csv", point_id))

    if (file.exists(output_file)) return(NULL)

    tryCatch({
      .download_era5_land_point_csv(
        latitude = latitude,
        longitude = longitude,
        start_date = start_date_str,
        end_date = end_date_str,
        target_file = raw_csv,
        cds_user = cds_user,
        verbose = FALSE
      )

      hourly <- .read_era5_land_hourly_csv(raw_csv)
      daily <- .aggregate_era5_land_to_daily(
        hourly_df = hourly,
        start_date = start_date_str,
        end_date = end_date_str,
        utc_offset_hours = utc_offset_hours
      )

      missing_before <- attr(daily, "missing_before")
      missing_after <- attr(daily, "missing_after")
      if (any(missing_before > 0)) {
        repair_msg <- paste(sprintf("%s %d->%d", names(missing_before), missing_before, missing_after), collapse = "; ")
        .log_worker_message(log_file, "WARN", point_id,
                            sprintf("Missing ERA5-Land daily values detected and repaired where possible: %s", repair_msg))
      }

      .write_dssat_weather_file(
        weather_data = daily,
        latitude = latitude,
        longitude = longitude,
        output_file = output_file,
        point_id = point_id
      )

      if (!keep_raw_downloads && file.exists(raw_csv)) unlink(raw_csv)
      .log_worker_message(log_file, "INFO", point_id,
                          sprintf("Successfully created ERA5-Land weather file: %s", basename(output_file)))
      NULL
    }, error = function(e) {
      .log_worker_message(log_file, "ERROR", point_id, conditionMessage(e))
      NULL
    })
  }

  requested_cores <- suppressWarnings(as.integer(n_cores))
  if (is.na(requested_cores) || requested_cores < 1L) requested_cores <- 1L
  if (requested_cores == 1L || length(ids) <= 1L) {
    # Keep one-core operation in the current process. Besides avoiding PSOCK
    # startup overhead, this makes debugging and mocked/offline tests reliable.
    lapply(seq_along(ids), .process_one)
  } else {
    cl <- parallel::makeCluster(min(requested_cores, length(ids)))
    doParallel::registerDoParallel(cl)
    on.exit(parallel::stopCluster(cl), add = TRUE)

    # Export required internal functions and package names to parallel workers.
    parallel::clusterExport(
      cl,
      varlist = c(
        ".download_era5_land_point_csv", ".build_era5_land_request",
        ".read_era5_land_hourly_csv", ".parse_hourly_datetime",
        ".find_first_matching_column", ".ensure_numeric",
        ".aggregate_era5_land_to_daily", ".fill_na_with_neighbor_mean",
        ".calc_rh_from_temp_dew", ".write_dssat_weather_file",
        ".log_worker_message", ".dssatutils_cds_default_url",
        ".dssatutils_cds_rc_candidates", ".dssatutils_read_cdsapirc",
        ".dssatutils_prompt_secret", "setup_cds_credentials",
        ".dssatutils_ensure_cds_credentials", "cds_user"
      ),
      envir = environment()
    )

    foreach::foreach(
      i = seq_along(ids),
      .packages = c("ecmwfr", "readr", "dplyr", "lubridate"),
      .export = ".process_one"
    ) %dopar% .process_one(i)
  }

  invisible(TRUE)
}
